src/Ratios.jl

```diff
@@ -32,8 +67,10 @@ Rational{T}(r::SimpleRatio{S}) where {T<:Integer, S<:Integer} = convert(T, r.num
 /(x::Integer, y::SimpleRatio) = SimpleRatio(x*y.den, y.num)
 +(x::Integer, y::SimpleRatio) = SimpleRatio(x*y.den + y.num, y.den)
 -(x::Integer, y::SimpleRatio) = SimpleRatio(x*y.den - y.num, y.den)
-+(x::SimpleRatio, y::SimpleRatio) = SimpleRatio(x.num*y.den + x.den*y.num, x.den*y.den)
--(x::SimpleRatio, y::SimpleRatio) = SimpleRatio(x.num*y.den - x.den*y.num, x.den*y.den)
++(x::SimpleRatio, y::SimpleRatio) = x.den == y.den ? SimpleRatio(x.num + y.num, x.den) :
+                                                     SimpleRatio(x.num*y.den + x.den*y.num, x.den*y.den)
+-(x::SimpleRatio, y::SimpleRatio) = x.den == y.den ? SimpleRatio(x.num - y.num, x.den) :
+                                                     SimpleRatio(x.num*y.den - x.den*y.num, x.den*y.den)
```

Comment:
The one other question I had with this implementation is if using `ifelse` might be advantageous over the ternary operator in order to avoid branching logic. However, I have not found a good test case to evaluate this. Obviously on a small scale, doing only one calculation is better. The question is if there is any vectorization that occurs before this change and if that vectorization is preserved by introducing this branch.

```suggestion
+(x::SimpleRatio, y::SimpleRatio) = ifelse(x.den == y.den, SimpleRatio(x.num + y.num, x.den),
                                                           SimpleRatio(x.num*y.den + x.den*y.num, x.den*y.den) )
-(x::SimpleRatio, y::SimpleRatio) = ifelse(x.den == y.den, SimpleRatio(x.num - y.num, x.den),
                                                           SimpleRatio(x.num*y.den - x.den*y.num, x.den*y.den) )
```

Reply:
Agreed in principle, but
```julia
julia> f(x, y) = x == y ? 1 : 1.0
f (generic function with 1 method)

julia> @code_llvm f(2, 3)
;  @ REPL[4]:1 within `f'
define { {}*, i8 } @julia_f_413([8 x i8]* noalias nocapture align 8 dereferenceable(8) %0, i64 signext %1, i64 signext %2) {
top:
; ┌ @ promotion.jl:410 within `=='
   %.not = icmp eq i64 %1, %2
; └
  %spec.select = select i1 %.not, { {}*, i8 } { {}* inttoptr (i64 140004302721120 to {}*), i8 -126 }, { {}*, i8 } { {}* inttoptr (i64 140004314812816 to {}*), i8 -127 }
  ret { {}*, i8 } %spec.select
}
```
`select` already avoids a branch and is SIMD-friendly.

Curiously, the LLVM for `ifelse` looks much worse:
```julia
julia> fifelse(x, y) = ifelse(x == y, 1, 1.0)
fifelse (generic function with 1 method)

julia> @code_llvm fifelse(2, 3)
;  @ REPL[6]:1 within `fifelse'
define { {}*, i8 } @julia_fifelse_435([8 x i8]* noalias nocapture align 8 dereferenceable(8) %0, i64 signext %1, i64 signext %2) {
top:
; ┌ @ promotion.jl:410 within `=='
   %.not = icmp eq i64 %1, %2
; └
  %3 = select i1 %.not, {}* inttoptr (i64 140004302721120 to {}*), {}* inttoptr (i64 140003489954752 to {}*)
  %4 = select i1 %.not, i1 icmp eq ({}* inttoptr (i64 140004302721120 to {}*), {}* null), i1 icmp eq ({}* inttoptr (i64 140003489954752 to {}*), {}* null)
  br i1 %4, label %postnull, label %nonnull

nonnull:                                          ; preds = %top
  %5 = bitcast {}* %3 to i64*
  %6 = getelementptr inbounds i64, i64* %5, i64 -1
  %7 = load atomic i64, i64* %6 unordered, align 8
  %8 = and i64 %7, -16
  %9 = inttoptr i64 %8 to {}*
  br label %postnull

postnull:                                         ; preds = %nonnull, %top
  %10 = phi {}* [ null, %top ], [ %9, %nonnull ]
  %11 = icmp eq {}* %10, inttoptr (i64 140004383587200 to {}*)
  %12 = zext i1 %11 to i8
  %13 = icmp eq {}* %10, inttoptr (i64 140004380990320 to {}*)
  %.op = or i8 %12, -128
  %14 = select i1 %13, i8 -126, i8 %.op
  %15 = insertvalue { {}*, i8 } undef, {}* %3, 0
  %16 = insertvalue { {}*, i8 } %15, i8 %14, 1
  ret { {}*, i8 } %16
}
```

I am puzzled, but it seems clear which one we want.

Reply:
I think it has to do with type stability.
```julia
julia> f(x, y) = x == y ? 1.0 : 2.0
f (generic function with 1 method)

julia> @code_llvm f(2,3)
;  @ REPL[23]:1 within `f`
; Function Attrs: uwtable
define double @julia_f_542(i64 signext %0, i64 signext %1) #0 {
top:
; ┌ @ promotion.jl:424 within `==`
   %.not = icmp eq i64 %0, %1
; └
  %spec.select = select i1 %.not, double 1.000000e+00, double 2.000000e+00
  ret double %spec.select
}

julia> fifelse(x, y) = ifelse(x == y, 1.0, 2.0)
fifelse (generic function with 1 method)

julia> @code_llvm fifelse(2,3)
;  @ REPL[25]:1 within `fifelse`
; Function Attrs: uwtable
define double @julia_fifelse_544(i64 signext %0, i64 signext %1) #0 {
top:
; ┌ @ promotion.jl:424 within `==`
   %.not = icmp eq i64 %0, %1
; └
  %2 = select i1 %.not, double 1.000000e+00, double 2.000000e+00
  ret double %2
}
```

The ternary operator seems generate better code in either case